File: sdk/javascript/src/indexes.ts
import type { RequestFn, IndexInfo, IndexCreatedResponse } from "./types.js"

export class IndexesModule {
  constructor(private readonly request: RequestFn) {}

  async list(): Promise<string[]> {
    const body = await this.request<{ indexes: string[] }>("GET", "/indexes")
    return body.indexes
  }

  async get(name: string): Promise<IndexInfo> {
    return this.request<IndexInfo>("GET", `/indexes/${encodeURIComponent(name)}`)
  }

  async create(name: string): Promise<IndexCreatedResponse> {
    return this.request<IndexCreatedResponse>("POST", "/indexes", { name })
  }

  async delete(name: string): Promise<void> {
    await this.request<void>("DELETE", `/indexes/${encodeURIComponent(name)}`)
  }
}

File: sdk/python/src/pelisearch/indexes.py
from __future__ import annotations

from typing import Callable

from .types import IndexCreatedResponse, IndexInfo


class IndexOperations:
    def __init__(self, request: Callable[..., object]) -> None:
        self._request = request

    def list_indexes(self) -> list[str]:
        body = self._request("GET", "/indexes")
        return body["indexes"]

    def get_index(self, name: str) -> IndexInfo:
        body = self._request("GET", f"/indexes/{_esc(name)}")
        return IndexInfo(**body)

    def create_index(self, name: str) -> IndexCreatedResponse:
        body = self._request("POST", "/indexes", {"name": name})
        return IndexCreatedResponse(**body)

    def delete_index(self, name: str) -> None:
        self._request("DELETE", f"/indexes/{_esc(name)}")


def _esc(s: str) -> str:
    import urllib.parse

    return urllib.parse.quote(s, safe="")
